spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -561,18 +564,21 @@ public void run() {
 
 		private void stopInvokerAndCommitManualAcks() {
 			this.invoker.stop();
+			this.listenerInvokerFuture.cancel(true);
```

Comment:
Thinking this through again - `cancel` doesn't feel right, as it deprives us from the option to wait for normal completion of the invoker runnable - the runnable will stop, but will do this in the background, potentially processing some messages before it stops and keeping to send acks.

I believe that the correct thing to do (albeit complicated) is to interrupt the thread that executes the invoker directly during stop, wait for normal completion on `get()` - the most likely outcome - and cancel if we are dealing with a non-cooperative listener (whose acks we can discard).